File: 438-find-all-anagrams-in-a-string/438-find-all-anagrams-in-a-string.py
class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        if len(p)>len(s): return []
        mp=[0]*26
        for i in p: mp[ord(i)-97]+=1
        
        j, ms = 0, [0]*26
        while j<len(p):
            ms[ord(s[j])-97]+=1
            j+=1
        
        i, j, ans = 0, j-1, []
        while j<len(s):
            if mp==ms: ans.append(i)
            ms[ord(s[i])-97]-=1
            i+=1
            j+=1
            if j<len(s): ms[ord(s[j])-97]+=1
        
        return ans            

File: 438-find-all-anagrams-in-a-string/438-find-all-anagrams-in-a-string.java
class Solution {
    public List<Integer> findAnagrams(String s, String p) {
        if(s.length()<p.length()) return new ArrayList<>();
        
        int[] mp=new int[26];
        for(int i=0; i<p.length(); i++){
            mp[p.charAt(i)-97]++;
        }
        
        int j=0;
        int[] ms = new int[26];
        while(j<p.length()) ms[s.charAt(j++)-97]++;
        
        j--;
        int i=0;
        List<Integer> ans=new ArrayList<>();
        while(j<s.length()){
            if(Arrays.equals(mp, ms)) ans.add(i);
            j++;
            if(j<s.length()) ms[s.charAt(j)-97]++;
            ms[s.charAt(i)-97]--;
            i++;
        }
        
        return ans;
    }
}